corehq/apps/hqadmin/tasks.py

```diff
@@ -201,3 +204,39 @@ def track_pg_limits():
                 cursor.execute(f'select last_value from "{sequence}"')
                 current_value = cursor.fetchone()[0]
                 metrics_gauge('commcare.postgres.sequence.current_value', current_value, {'table': table, 'database': db})
+
+
+@periodic_task(queue='background_queue', run_every=crontab(minute="0", hour="4"))
+def reconcile_es_cases():
+    _reconcile_es_data('case', 'commcare.elasticsearch.stale_cases', 'reconcile_es_cases')
+
+
+@periodic_task(queue='background_queue', run_every=crontab(minute="0", hour="4"))
+def reconcile_es_forms():
+    _reconcile_es_data('form', 'commcare.elasticsearch.stale_forms', 'reconcile_es_forms')
+
+
+def _reconcile_es_data(data_type, metric, blob_parent_id):
+    today = date.today()
+    two_days_ago = today - timedelta(days=2)
+    with NamedTemporaryFile('w', delete=False) as output_file:
+        file_name = output_file.name
+        call_command('stale_data_in_es', data_type, start=two_days_ago.isoformat(), stdout=output_file)
+    with open(file_name, 'r') as f:
+        line_count = sum(1 for line in f)
+        # ignore the headers
+        stale_docs = line_count - 1
+        metrics_gauge(metric, stale_docs)
+    call_command('republish_doc_changes', file_name)
```

Comment:
I'm concerned that automatically "fixing" the issue will cause us to become complacent about bugs in other places. Especially since this blows away state of the stale doc, which might be useful evidence. It seems like it would be better to track down the places that are causing stale docs in ES.

Reply:
I agree we want to preserve a reasonable amount of evidence to let us look into problems, but I think we've found this leaky pipeline problem to be obstinately persistent over many years. Not doing a sweep because it felt like admitting defeat is sort of the approach we've been taking that whole time, and so instead what we've done is large sweeps when we get support tickets. I think this PR puts us on a much better path: it surfaces all gaps in near real time (i.e on a day scale) and graphs them for us so we can tell whether our problem is a low-level steady leak or big bursts during specific incidents, while at the same time fixing the problems it finds so we're not waiting for customers to report them. It's both better for the customer than the status quo, and gives us more ability to identify leaks, as we'll have not just a graph of the daily number, but also the log output of that process that Cal's storing in blobdb. You mentioned another piece of evidence that could be useful, namely the full doc content before it's fixed. That's a fairly simple thing to add (in a verbose mode for the logging process perhaps) if digging into the ids doesn't get us anywhere. I think what we don't want to do is keep things broken for customers for the sake of preserving evidence; instead I'd advocate for this approach, of having the automated sweep process preserve the evidence that will be most directly useful to investigation, and then iterate on what evidence we're saving to answer the questions that come up during investigation.

Reply:
I see some value in this, but I still thinks it encourages us to work in a sub-optimal state where we have added complexity in our system to tape over holes that should be fixed properly. Also, I think it is an incorrect way to structure a data pipeline at the core of our product.

> we've found this leaky pipeline problem to be obstinately persistent over many years

I agree, this has been a long-standing issue. I think we are not putting enough effort into fixing the real root causes of bugs when they come up. Case in point is the most recent occurrence of stale cases in ES. My understanding is that we came up with a theory about what caused the problem without substantiating that theory. Then went ahead and refreshed the stale cases (blowing away evidence and state that may have been useful in determining the root cause) and then monitored to make sure no stale cases recurred in that set. It appears we are adopting this PR as part of the "fix" for that issue, even though a different issue definitely caused the problem and we are not putting effort into finding that cause.

Part of the problem is that in some ways we treat ES as a non-primary data source (we're not too worried if its out of sync from time to time), but in other ways we treat it as a primary data source (many reports use it as a source of truth). I can imagine a future scenario where a customer complains of a report showing incorrect data, and then it is magically fixed and cannot be reproduced by the time we get around to looking at it. Judging from past behavior, I would bet that we look into the issue, realize it could have been caused by stale docs in ES, then remember we have an automated fix for that (this PR), and move on without finding the real cause. To me that looks like a fundamentally broken data pipeline because it cannot be trusted to accurately return data that has been updated within the past day.

Reply:
In addition to all all that Danny said about how this puts us in a much better position to understand the extent and cause of any data consistency issues than we currently are, including recording significantly more information than we currently have, I want to add a little more of the rationale for this work.

This pull request is another step in the direction of a more resilient system, just like the retry code that exists at every step of the form submission and processing pipeline. Building a system that can still get data into reports in a timely manner,  in the face of errors should be seen as a positive good, not a bandaid. People rely on this data to be right, and so we should do our best to make sure that is the case, even in the face of inevitable errors.

> I'm concerned that automatically "fixing" the issue will cause us to become complacent about bugs in other places

It is definitely true that given this PR, we may decide that certain issues are no longer worth investigating, that a certain class or level of pillow error is acceptable if we know that it will be resolved within 24 hours, but it also puts us in a much better position to understand what is going on, to see exactly where stale data occurs and with what frequency. It will allow us to make more informed decisions about when and how we investigate issues in ES data, without sacrificing user experience or trust. 

> blowing away evidence and state that may have been useful

On the contrary, I think this PR collects a significant amount of evidence about the stale documents, far more than we had in the past, and to the extent document state contributes to staleness, it is the state of the document in the primary database (issues like the too large fields) that cause those errors, not the previous version of the doc held in ES. That isn't the stale data or where the error occurred, and this PR does not change the state of the data in postgres, although does record enough info to reconstruct it. Furthermore, for any case that simply resaving (as this PR does) is successful, it could not have been the state of the doc that caused the error since without changing the state, a pillow was able to successfully save it.

> Part of the problem is that in some ways we treat ES as a non-primary data source (we're not too worried if its out of sync from time to time),

I don't think this is true anymore, if it was previously. We have an obligation to our customers who are using these reports (and related features like case search) to inform real decisions based on real data about real people to have these reports be as correct as possible. If we have the ability to correct issues in the data, choosing not to while we attempt to track down the issue with each and every individual document (many of which will end up being things like transient network or cluster health issues, that we will never be able to find), punishes our customers customers, and is an inefficient use of limited developer resources. Our customers do not care how the data gets there, and whether it was the first or second attempt, they care whether the data is correct.

> To me that looks like a fundamentally broken data pipeline because it cannot be trusted to accurately return data that has been updated within the past day.

This may be true, although we will be better able to answer that question after this PR than before it. But if it is true, I still think this is a necessary step to fixing the pipeline. If it is the case that our ES pipeline is broken and untrustworthy, we need a way to make sure the data in it is correct, while we work on the larger, necessary fixes.

Reply:
I am not arguing at all that we should not implement the parts of this PR that are automatically _detecting and recording_ occurrences of stale documents in ES. That is of undisputed value. The only part of this that I have a problem with is this single line that republishes changes on stale documents.

It will probably be more productive to move this discussion to voice rather than tossing more walls of text back and forth here.